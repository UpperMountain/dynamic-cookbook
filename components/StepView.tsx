import React from "react";
import { View, Text, TouchableOpacity } from "react-native";
import { Step } from "../lib/graph";
import LeftLine from "./LeftLine";
import Markdown from "react-native-markdown-renderer";
import Padded from "./Padded";
import Card from "./Card";
import { padding } from "../lib/theme";
import { MaterialIcons } from "@expo/vector-icons";

const topSize = 40;

interface CircledProps {
  children: string;
}
function Circled(props: CircledProps) {
  const { children } = props;
  return (
    <View
      style={{
        width: topSize,
        height: topSize,
        display: "flex",
        justifyContent: "center",
        alignItems: "center",
        borderWidth: 1,
        borderColor: "black",
        borderRadius: 25
      }}
    >
      <Text style={{ fontSize: 18 }}>{children}</Text>
    </View>
  );
}

function stringFormatInterval(seconds: number) {
  let minutes: number | null = null;
  let hours: number | null = null;
  if (seconds > 60) {
    minutes = Math.floor(seconds / 60);
    seconds = seconds % 60;
  }
  if (minutes && minutes > 60) {
    hours = Math.floor(minutes / 60);
    minutes = minutes % 60;
  }

  let out = "";
  if (hours) out += hours.toString().padStart(2, "0") + ":";
  if (minutes) out += minutes.toString().padStart(2, "0") + ":";
  out += seconds.toString().padStart(2, "0");

  if (!hours && !minutes) {
    out += " s";
  }

  return out;
}

export interface StepActionProps {
  // Time remaining, or undefined if not started
  remaining?: number;

  // if this card is, or leads to, a timer
  timer?: boolean;

  // "The onions are chopped"
  until: string;

  // tap callback
  onPress: () => void;
}
export function StepAction(props: StepActionProps) {
  const { timer = false, remaining = null, until, onPress } = props;
  if (remaining !== null && !timer) {
    throw "StepAction: passed time remaining without timer enabled";
  }
  return (
    <Card
      shadowAmt={0.5}
      innerStyle={{
        marginLeft: padding / 2,
        marginRight: padding / 2,
        borderRadius: 10
      }}
    >
      <TouchableOpacity
        onPress={onPress}
        style={{ marginLeft: -padding / 2, marginRight: -padding / 2 }}
      >
        <LeftLine
          hidden
          aside={
            <View
              style={{
                flex: 1,
                display: "flex",
                flexDirection: "column",
                justifyContent: "center"
              }}
            >
              <MaterialIcons
                name={
                  timer
                    ? remaining != null
                      ? "timer"
                      : "play-circle-outline"
                    : "arrow-forward"
                }
                size={40}
              />
            </View>
          }
        >
          <Padded right vertical>
            {remaining != null && (
              <Text style={{ fontSize: 30 }}>
                {stringFormatInterval(remaining)}
              </Text>
            )}
            <View
              style={{
                display: "flex",
                flexDirection: "row",
                alignItems: "center"
              }}
            >
              {remaining != null && (
                <MaterialIcons
                  name="arrow-forward"
                  size={20}
                  style={{ marginRight: 5 }}
                />
              )}
              <Text style={{ fontSize: 20 }}>{until}</Text>
            </View>
          </Padded>
        </LeftLine>
      </TouchableOpacity>
    </Card>
  );
}

export interface Props {
  step: Step;
  num: number;
}

export default function StepView(props: Props) {
  const { step, num } = props;
  return (
    <LeftLine aside={<Circled>{num.toString()}</Circled>}>
      <Padded right bottom>
        <View
          style={{
            minHeight: topSize,
            display: "flex",
            flexDirection: "column",
            justifyContent: "center"
          }}
        >
          <Text style={{ fontSize: 20 }}>{step.name}</Text>
        </View>
        {step.body ? (
          <Markdown>{step.body}</Markdown>
        ) : (
          <View style={{ height: padding }} />
        )}
      </Padded>
    </LeftLine>
  );
}
