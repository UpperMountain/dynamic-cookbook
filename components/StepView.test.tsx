import React from "react";
import StepView, { StepAction } from "./StepView";
import renderer from "react-test-renderer";

describe("<StepView/>", () => {
  it("should render without exploding", () => {
    renderer
      .create(
        <StepView
          step={{
            kind: "step",
            until: "[until text]",
            name: "[step name]",
            requires: []
          }}
          num={10}
        />
      )
      .unmount();
  });

  it("should pass snapshot tests", () => {
    const cmp = renderer.create(
      <>
        <StepView
          num={1}
          step={{
            kind: "step",
            until: "until 1",
            name: "name 1",
            body: "body 1",
            requires: []
          }}
        />
        <StepView
          num={2}
          step={{
            kind: "step",
            until: "until 2",
            name: "name 2",
            requires: []
          }}
        />
      </>
    );
    expect(cmp.toJSON()).toMatchSnapshot();
    cmp.unmount();
  });
});

describe("<StepAction/>", () => {
  it("should render without exploding", () => {
    renderer
      .create(<StepAction until="until text" onPress={jest.fn()} />)
      .unmount();
  });

  it("should pass snapshot tests", () => {
    const cmp = renderer.create(
      <>
        <StepAction until="until text" onPress={jest.fn()} />
        <StepAction
          timer={true}
          remaining={10000}
          until="until text"
          onPress={jest.fn()}
        />
        <StepAction
          timer={true}
          remaining={100}
          until="until text"
          onPress={jest.fn()}
        />
        <StepAction
          timer={true}
          remaining={10}
          until="until text"
          onPress={jest.fn()}
        />
        <StepAction timer={true} until="until text" onPress={jest.fn()} />
      </>
    );
    expect(cmp.toJSON()).toMatchSnapshot();
    cmp.unmount();
  });

  it("should throw if props make no sense", () => {
    // don't print the react error
    jest.spyOn(global.console, "error").mockImplementation(() => jest.fn());

    expect(() => {
      renderer
        .create(
          <StepAction remaining={1} until="until text" onPress={jest.fn()} />
        )
        .unmount();
    }).toThrow();

    // @ts-ignore
    console.error.mockClear();
  });
});
