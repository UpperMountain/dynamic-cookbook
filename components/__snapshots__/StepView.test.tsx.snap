// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StepView/> should pass snapshot tests 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "display": "flex",
          "flexDirection": "row",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "width": 75,
          },
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "borderColor": "black",
              "borderRadius": 25,
              "borderWidth": 1,
              "display": "flex",
              "height": 40,
              "justifyContent": "center",
              "width": 40,
            },
            undefined,
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 17,
              },
              undefined,
            ]
          }
        >
          1
        </Text>
      </View>
      <View
        style={
          Object {
            "borderLeftColor": "black",
            "borderLeftWidth": 1,
            "flex": 1,
          }
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "marginLeft": 0,
          },
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "paddingBottom": 20,
              "paddingLeft": 0,
              "paddingRight": 20,
              "paddingTop": 0,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Object {
              "display": "flex",
              "flexDirection": "column",
              "justifyContent": "center",
              "minHeight": 40,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 22,
                },
                undefined,
              ]
            }
          >
            name 1
          </Text>
        </View>
        <View
          style={Object {}}
        >
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "flexDirection": "row",
                "flexWrap": "wrap",
                "justifyContent": "flex-start",
                "marginBottom": 10,
                "marginTop": 10,
              }
            }
          >
            <Text
              style={Object {}}
            >
              <Text>
                body 1
              </Text>
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "display": "flex",
          "flexDirection": "row",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "width": 75,
          },
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "borderColor": "black",
              "borderRadius": 25,
              "borderWidth": 1,
              "display": "flex",
              "height": 40,
              "justifyContent": "center",
              "width": 40,
            },
            undefined,
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 17,
              },
              undefined,
            ]
          }
        >
          2
        </Text>
      </View>
      <View
        style={
          Object {
            "borderLeftColor": "black",
            "borderLeftWidth": 1,
            "flex": 1,
          }
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "marginLeft": 0,
          },
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "paddingBottom": 20,
              "paddingLeft": 0,
              "paddingRight": 20,
              "paddingTop": 0,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Object {
              "display": "flex",
              "flexDirection": "column",
              "justifyContent": "center",
              "minHeight": 40,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 22,
                },
                undefined,
              ]
            }
          >
            name 2
          </Text>
        </View>
        <View
          style={
            Object {
              "height": 20,
            }
          }
        />
      </View>
    </View>
  </View>,
]
`;
